### PR TITLE
auth: Implementation fo AzureSub provider that leverages federated credentials

### PR DESCRIPTION
**Pull request by nturinski**
I've explained this quite a bit offline, but just some context for the PR.

We are using federated credentials (service principals) in order to log in to Azure. This token is used in place of VS Code sessions and passed to our subscription client.  The resources extension will need to swap which sub provider it is using based on whether or not it is in a testing environment, but there should be no code changes beyond that.

Still need to add the README. We should translate Hossam's document from OneNote to markdown and include it as a README.

### CODE REVIEW

File: auth/package.json [hunk 1]
```diff
@@ -32,6 +32,7 @@
     },
     "devDependencies": {
         "@azure/core-auth": "^1.4.0",
+        "@azure/identity": "^4.2.0",
```
Comment (bwateratmsft) [inline auth/package.json:35]: I don't see this package being used

File: auth/src/AzureDevOpsSubscriptionProvider.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        return !!this._tokenCredential;
+    }
+
+    public async signOut(): Promise<void> {
+        this._tokenCredential = undefined;
+    }
+
+    public async getTenants(): Promise<TenantIdDescription[]> {
+        return [{
+            tenantId: this._tokenCredential?.tenantId,
```
Comment (alexweininger) [inline auth/src/AzureDevOpsSubscriptionProvider.ts:94]: So there can only ever be one tenant associated with the token? I also don't see where the `tenantId` property is ever set.

**Review by alexweininger — APPROVED** [review]

File: auth/package.json (hunk 1 above)
Comment (nturinski) [reply auth/package.json:35]: https://github.com/microsoft/vscode-azuretools/pull/1723/files#diff-9c30a2d74591058fcf9d06dbc7074281071ad805af0619e472a83df969ac8284R202

That lazy load tho~~

Comment (nturinski) [reply auth/package.json:35]: Note about the new dependencies-- they were already included in the `@azure/ms-rest-azure-env` but I didn't want to import implicitly, so just included them in the package.json. It shouldn't be introducing new dependencies.

File: auth/src/AzureDevOpsSubscriptionProvider.ts [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                        scopes: scopes || [],
+                    };
+
+                }
+            }
+        };
+    }
+
+    public onDidSignIn: Event<void> = () => { return new Disposable(() => { /*empty*/ }) };
+    public onDidSignOut: Event<void> = () => { return new Disposable(() => { /*empty*/ }) };
```
Comment (alexweininger) [inline auth/src/AzureDevOpsSubscriptionProvider.ts:163]: You could probably add support for these pretty easily

**Review by alexweininger — APPROVED** [review]

File: auth/src/AzureDevOpsSubscriptionProvider.ts [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            process.env.SYSTEM_TEAMFOUNDATIONCOLLECTIONURI: ${process.env.SYSTEM_TEAMFOUNDATIONCOLLECTIONURI ? "✅" : "❌"}\n
+            process.env.SYSTEM_TEAMPROJECTID: ${process.env.SYSTEM_TEAMPROJECTID ? "✅" : "❌"}\n
+            process.env.SYSTEM_PLANID: ${process.env.SYSTEM_PLANID ? "✅" : "❌"}\n
+            process.env.SYSTEM_JOBID: ${process.env.SYSTEM_JOBID ? "✅" : "❌"}\n
+            REMEMBER: process.env.SYSTEM_ACCESSTOKEN must be explicitly mapped!\n
+            https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#systemaccesstoken
+        `);
+        }
+
+        const oidcRequestUrl = `${teamFoundationCollectionUri}${teamProjectId}/_apis/distributedtask/hubs/build/plans/${planId}/jobs/${jobId}/oidctoken?api-version=7.1-preview.1&serviceConnectionId=${serviceConnectionId}`;
```
Comment (MicroFish91) [inline auth/src/AzureDevOpsSubscriptionProvider.ts:199]: Question: How did you figure out the exact format for the `oidcRequestUrl`?  Was this referenced in the MS Learn docs somewhere or by manually inspecting a job that was spun up?

Comment (MicroFish91) [reply auth/src/AzureDevOpsSubscriptionProvider.ts:199]: Nevermind, disregard I'm assuming it's the same format as this :)
https://learn.microsoft.com/en-us/rest/api/azure/devops/distributedtask/oidctoken/create

Comment (alexweininger) [reply auth/src/AzureDevOpsSubscriptionProvider.ts:199]: Good question. We might want to add that link to a comment above this.

**Pull request merged**